sled-agent/src/instance.rs

```diff
@@ -382,6 +383,53 @@ impl InstanceMonitorRunner {
             Err(e) if self.tx_monitor.is_closed() => {
                 Err(BackoffError::permanent(e))
             }
+            // If we couldn't communicate with propolis-server, let's make sure
+            // the zone is still there...
+            Err(e @ PropolisClientError::CommunicationError(_)) => {
+                match Zones::find(&self.zone_name).await {
+                    Ok(None) => {
+                        // Oh it's GONE!
+                        info!(
+                            self.log,
+                            "Propolis zone is Way Gone!";
+                            "zone" => %self.zone_name,
+                        );
+                        Ok(InstanceMonitorUpdate::ZoneGone)
+                    }
+                    Ok(Some(zone)) if zone.state() == zone::State::Running => {
+                        warn!(
+                            self.log,
+                            "communication error checking up on Propolis, but \
+                             the zone is still running...";
+                            "error" => %e,
+                            "zone" => %self.zone_name,
+                        );
+                        Err(BackoffError::transient(e))
+                    }
+                    Ok(Some(zone)) => {
+                        info!(
+                            self.log,
+                            "Propolis zone is no longer running!";
+                            "error" => %e,
+                            "zone" => %self.zone_name,
+                            "zone_state" => ?zone.state(),
+                        );
+                        Ok(InstanceMonitorUpdate::ZoneGone)
+                    }
```

Comment:
I believe this is already true - but to confirm, `monitor` can only be called after the VMM's zone was initialized, correct?

(e.g., there's no concern about racing with zone startup here, right?)

Reply:
That's correct, the function that starts the state monitor task is called with a `RunningZone`, which is only returned if the zone has already reached running.